File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        if (n==0 || N==0){
            return 1.0;
        }
    
        if (N < 0) {
            N = -N;
            x = 1 / x;
        }
        if(N%2==0){
            double t=myPow(x,(int) (N/2));
            return(t*t);
        }
        else{
            double t=myPow(x,(int)(N-1));
            return(x*t);
        }  
    }
}

File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        long long m = n;
        if(m<0){
            m=m*-1;
        }
        long double ans = 1.0;
        while(m>0){
            if(m%2==1){
                ans = ans * x;
                m=m-1;
            }
            else{
                m=m/2;
                x = x*x;
            }
        }
        if(n>0){
            return ans;
        }
        else{
            ans = 1.0/ans;
            return ans;
        }
    }
};